File: examples/integrators/with_jacobian.py
# /// script
# dependencies = ["brahe", "numpy"]
# ///
"""
Demonstrates integration with state transition matrix propagation.

This example shows how to propagate variational equations using a numerical
Jacobian for orbit determination applications.
"""

import brahe as bh
import numpy as np

# Initialize EOP
bh.initialize_eop()


# Define two-body dynamics
def dynamics(t, state):
    mu = bh.GM_EARTH
    r = state[0:3]
    v = state[3:6]
    r_norm = np.linalg.norm(r)
    a = -mu / r_norm**3 * r
    return np.concatenate([v, a])


# Create numerical Jacobian for variational equations
jacobian = bh.NumericalJacobian.central(dynamics).with_adaptive(
    scale_factor=1.0, min_value=1e-6
)

# Initial orbit (LEO)
r0 = np.array([bh.R_EARTH + 600e3, 0.0, 0.0])
v0 = np.array([0.0, 7.5e3, 0.0])
state0 = np.concatenate([r0, v0])

# Initial state transition matrix (identity)
phi0 = np.eye(6)

print("Integration with State Transition Matrix")
print(f"Initial orbit: {r0[0] / 1e3:.1f} km altitude")

# Create integrator with Jacobian
config = bh.IntegratorConfig.adaptive(abs_tol=1e-12, rel_tol=1e-11)
integrator = bh.DP54Integrator(6, dynamics, jacobian=jacobian, config=config)

# Propagate for one orbit period
t = 0.0
state = state0.copy()
phi = phi0.copy()
dt = 60.0

# Approximate orbital period
period = bh.orbital_period(np.linalg.norm(r0))

print("Time   Position STM[0,0]  Velocity STM[3,3]  Det(STM)")
print("-" * 60)

steps = 0
while t < period:
    # Propagate state and STM together (adaptive integrator returns 5-tuple)
    new_state, new_phi, dt_used, error_est, dt_next = integrator.step_with_varmat(
        t, state, phi, min(dt, period - t)
    )

    t += dt_used
    state = new_state
    phi = new_phi
    dt = dt_next
    steps += 1

    # Print progress
    if steps % 20 == 1:
        det_phi = np.linalg.det(phi)
        print(
            f"{t:6.0f}s    {phi[0, 0]:8.4f}      {phi[3, 3]:8.4f}        {det_phi:8.4f}"
        )

print(f"\nPropagation complete! ({steps} steps)")

# Example: Map initial position uncertainty to final uncertainty
print("\nExample: Uncertainty Propagation")
dx = 100.0
print(f"Initial position uncertainty: ±{dx} m in each direction")
delta_r0 = np.array([dx, dx, dx, 0.0, 0.0, 0.0])
delta_rf = phi @ delta_r0
print(
    f"Final position uncertainty: [{delta_rf[0]:.1f}, {delta_rf[1]:.1f}, {delta_rf[2]:.1f}] m"
)
print(
    f"Uncertainty growth: {np.linalg.norm(delta_rf[0:3]) / np.linalg.norm(delta_r0[0:3]):.1f}x"
)

File: examples/integrators/with_jacobian.rs
//! Demonstrates integration with state transition matrix propagation.

#[allow(unused_imports)]
use brahe::*;
use nalgebra::{DMatrix, DVector};

fn main() {
    // Initialize EOP
    initialize_eop().unwrap();

    // Define two-body dynamics (for integrator - takes optional params)
    let dynamics = |_t: f64, state: &DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        let r = nalgebra::Vector3::new(state[0], state[1], state[2]);
        let v = nalgebra::Vector3::new(state[3], state[4], state[5]);
        let r_norm = r.norm();
        let a = -GM_EARTH / r_norm.powi(3) * r;
        DVector::from_vec(vec![v[0], v[1], v[2], a[0], a[1], a[2]])
    };

    // Same dynamics for Jacobian (without params argument)
    let dynamics_for_jac = |_t: f64, state: &DVector<f64>, _params: Option<&DVector<f64>>| -> DVector<f64> {
        let r = nalgebra::Vector3::new(state[0], state[1], state[2]);
        let v = nalgebra::Vector3::new(state[3], state[4], state[5]);
        let r_norm = r.norm();
        let a = -GM_EARTH / r_norm.powi(3) * r;
        DVector::from_vec(vec![v[0], v[1], v[2], a[0], a[1], a[2]])
    };

    // Create numerical Jacobian for variational equations
    let jacobian = DNumericalJacobian::central(Box::new(dynamics_for_jac))
        .with_adaptive(1.0, 1e-6);

    // Initial orbit (LEO)
    let r0 = DVector::from_vec(vec![R_EARTH + 600e3, 0.0, 0.0]);
    let v0 = DVector::from_vec(vec![0.0, 7.5e3, 0.0]);
    let state0 = DVector::from_vec(vec![r0[0], r0[1], r0[2], v0[0], v0[1], v0[2]]);

    // Initial state transition matrix (identity)
    let phi0 = DMatrix::identity(6, 6);

    println!("Integration with State Transition Matrix");
    println!("Initial orbit: {:.1} km altitude", r0[0] / 1e3);

    // Create integrator with Jacobian
    let config = IntegratorConfig::adaptive(1e-12, 1e-11);
    let integrator = DormandPrince54DIntegrator::with_config(6, Box::new(dynamics), Some(Box::new(jacobian)), None, None, config);

    // Propagate for one orbit period
    let mut t = 0.0;
    let mut state = state0.clone();
    let mut phi = phi0.clone();
    let mut dt: f64 = 60.0;

    // Approximate orbital period
    let period = orbital_period(r0.norm());

    println!("Time   Position STM[0,0]  Velocity STM[3,3]  Det(STM)");
    println!("{}", "-".repeat(60));

    let mut steps = 0;
    while t < period {
        // Propagate state and STM together
        let result = integrator.step_with_varmat(t, state, None, phi, Some(dt.min(period - t)));

        let new_state = result.state;
        let new_phi = result.phi.unwrap();
        let dt_used = result.dt_used;
        let dt_next = result.dt_next;

        t += dt_used;
        state = new_state;
        phi = new_phi;
        dt = dt_next;
        steps += 1;

        // Print progress
        if steps % 20 == 1 {
            let det_phi = phi.clone().lu().determinant();
            println!("{:6.0}s    {:8.4}      {:8.4}        {:8.4}",
                     t, phi[(0, 0)], phi[(3, 3)], det_phi);
        }
    }

    println!("\nPropagation complete! ({} steps)", steps);

    // Example: Map initial position uncertainty to final uncertainty
    println!("\nExample: Uncertainty Propagation");
    let dx = 100.0;
    println!("Initial position uncertainty: ±{} m in each direction", dx);
    let delta_r0 = DVector::from_vec(vec![dx, dx, dx, 0.0, 0.0, 0.0]);
    let delta_rf = &phi * &delta_r0;
    println!("Final position uncertainty: [{:.1}, {:.1}, {:.1}] m",
             delta_rf[0], delta_rf[1], delta_rf[2]);

    let r0_norm = (delta_r0[0].powi(2) + delta_r0[1].powi(2) + delta_r0[2].powi(2)).sqrt();
    let rf_norm = (delta_rf[0].powi(2) + delta_rf[1].powi(2) + delta_rf[2].powi(2)).sqrt();
    println!("Uncertainty growth: {:.1}x", rf_norm / r0_norm);
}
